data/src/main/java/com/linkedin/data/codec/entitystream/AbstractDataDecoder.java

```diff
@@ -0,0 +1,369 @@
+              _expectedTokens = VALUE;
+            }
+            else
+            {
+              addValue(_parser.getString());
+            }
+            break;
+          case RAW_BYTES:
+            addValue(_parser.getRawBytes());
+            break;
+          case INTEGER:
+            addValue(_parser.getIntValue());
+            break;
+          case LONG:
+            addValue(_parser.getLongValue());
+            break;
+          case FLOAT:
+            addValue(_parser.getFloatValue());
+            break;
+          case DOUBLE:
+            addValue(_parser.getDoubleValue());
+            break;
+          case BOOL_TRUE:
+            addValue(Boolean.TRUE);
+            break;
+          case BOOL_FALSE:
+            addValue(Boolean.FALSE);
+            break;
+          case NULL:
+            addValue(Data.NULL);
+            break;
+          case NOT_AVAILABLE:
+            readNextChunk();
+            return;
+          default:
+            handleException(new Exception("Unexpected token " + token + " from data parser"));
+        }
+      }
+    }
+    catch (IOException e)
+    {
+      handleException(e);
+    }
+  }
+
+  /**
+   * Interface to new complex object, invoked at the start of parsing object.
+   */
+  protected abstract DataComplex createDataObject(NonBlockingDataParser parser);
+
+  /**
+   * Interface to new complex list, invoked at start of parsing array.
+   */
+  protected abstract DataComplex createDataList(NonBlockingDataParser parser);
+
+  protected final boolean isCurrList()
+  {
+    return _isCurrList;
+  }
+
+  protected final void validate(NonBlockingDataParser.Token token)
+  {
+    if (!(token == NOT_AVAILABLE || _expectedTokens.contains(token)))
+    {
+      handleException(new Exception("Expecting " + _expectedTokens + " but got " + token));
+    }
+  }
+
+  private void push(DataComplex dataComplex, boolean isList)
+  {
+    if (!(_isCurrList || _stack.isEmpty()))
+    {
+      _currFieldStack.push(_currField);
+      _currField = null;
+    }
+    _stack.push(dataComplex);
+    _isCurrList = isList;
+    updateExpected();
+  }
+
+  @SuppressWarnings("unchecked")
+  private void pop()
+  {
+    // The stack should never be empty because of token validation.
+    assert !_stack.isEmpty() : "Trying to pop empty stack";
+
+    DataComplex tmp = _stack.pop();
+    tmp = postProcessDataObject(tmp);
+    if (_stack.isEmpty())
+    {
+      _result = (T) tmp;
+      // No more tokens is expected.
+      _expectedTokens = NONE;
+    }
+    else
+    {
+      _isCurrList = _stack.peek() instanceof DataList;
+      if (!_isCurrList)
+      {
+        _currField = _currFieldStack.pop();
+      }
+      addValue(tmp);
+      updateExpected();
+    }
+  }
+
+  /**
+   * method invoked to do any post processing on complex object/list after its completely parsed and poped from stack
+   */
+  protected DataComplex postProcessDataObject(DataComplex dataComplex)
```

Comment:
nit: Method invoked.. 
s/poped/ popped

and the method should be named postProcessDataComplex? as DataObject is used for Maps in other places of this class.